### ISSUE TITLE
Chromecast fails to play when unsupported Audio Codec/Audio Channel's

### ISSUE THREAD

**Issue by Matt07211** [jellyfin/jellyfin#1312]
**Describe the bug**
When Chromecasting videos, if a video has an unsupported audio codec and/or an unsupported amount of audio channels, an ffmpeg process should be kicked off to make the stream compatible, this doesn't always happen, and leads to a grey screen when casting.

Example of it affecting other users: https://www.reddit.com/r/jellyfin/comments/bgqcxd/1031beta_9_android_chromecast_playback_broken_and/

**To Reproduce**
1. Connect to Chromecast
2. Go to an incompatible video and cast it to the Chromecast. Examples below
3. See Gray screen.

Example of a working direct stream: 
```
matt@thinkpod /mnt/3TB-Drive-01/Videos/Anime/Parasite Dolls/Specials $ mediainfo Parasite\ Dolls\ -\ SP\ 01\ -\ Get\ on\ the\ Beat\ Music\ Video.mkv 
General
Unique ID                                : 243534664841476153375476089322437287299 (0xB73712028D7C89778C2F31B326571D83)
Complete name                            : Parasite Dolls - SP 01 - Get on the Beat Music Video.mkv
Format                                   : Matroska
Format version                           : Version 2
File size                                : 272 MiB
Duration                                 : 4 min 39 s
Overall bit rate mode                    : Variable
Overall bit rate                         : 8 171 kb/s
Encoded date                             : UTC 2015-10-26 05:20:13
Writing application                      : MakeMKV v1.9.7 win(x86-release)
Writing library                          : libmakemkv v1.9.7 (1.3.0/1.4.1) win(x86-release)

Video
ID                                       : 1
Format                                   : MPEG Video
Format version                           : Version 2
Format profile                           : Main@Main
Format settings                          : CustomMatrix / BVOP
Format settings, BVOP                    : Yes
Format settings, Matrix                  : Custom
Format settings, GOP                     : M=3, N=12
Format settings, picture structure       : Frame
Codec ID                                 : V_MPEG2
Codec ID/Info                            : MPEG 1 or 2 Video
Duration                                 : 4 min 39 s
Bit rate mode                            : Variable
Bit rate                                 : 7 561 kb/s
Maximum bit rate                         : 8 900 kb/s
Width                                    : 720 pixels
Height                                   : 480 pixels
Display aspect ratio                     : 4:3
Frame rate mode                          : Constant
Frame rate                               : 29.970 (30000/1001) FPS
Standard                                 : NTSC
Color space                              : YUV
Chroma subsampling                       : 4:2:0
Bit depth                                : 8 bits
Scan type                                : Interlaced
Scan order                               : Top Field First
Compression mode                         : Lossy
Bits/(Pixel*Frame)                       : 0.730
Time code of first frame                 : 00:59:59:20
Time code source                         : Group of pictures header
GOP, Open/Closed                         : Open
GOP, Open/Closed of first frame          : Closed
Stream size                              : 252 MiB (93%)
Language                                 : English
Default                                  : No
Forced                                   : No

Audio #1
ID                                       : 2
Format                                   : AC-3
Format/Info                              : Audio Coding 3
Commercial name                          : Dolby Digital
Codec ID                                 : A_AC3
Duration                                 : 4 min 39 s
Bit rate mode                            : Constant
Bit rate                                 : 224 kb/s
Channel(s)                               : 2 channels
Channel layout                           : L R
Sampling rate                            : 48.0 kHz
Frame rate                               : 31.250 FPS (1536 SPF)
Bit depth                                : 16 bits
Compression mode                         : Lossy
Stream size                              : 7.47 MiB (3%)
Title                                    : Stereo
Language                                 : English
Service kind                             : Complete Main
Default                                  : Yes
Forced                                   : No

Audio #2
ID                                       : 3
Format                                   : AC-3
Format/Info                              : Audio Coding 3
Commercial name                          : Dolby Digital
Codec ID                                 : A_AC3
Duration                                 : 4 min 39 s
Bit rate mode                            : Constant
Bit rate                                 : 224 kb/s
Channel(s)                               : 2 channels
Channel layout                           : L R
Sampling rate                            : 48.0 kHz
Frame rate                               : 31.250 FPS (1536 SPF)
Bit depth                                : 16 bits
Compression mode                         : Lossy
Stream size                              : 7.47 MiB (3%)
Title                                    : Stereo
Language                                 : English
Service kind                             : Complete Main
Default                                  : No
Forced                                   : No

Menu
00:00:00.066                             : en:Chapter 01
00:04:38.745                             : en:Chapter 02
```
And an similar example with the only difference being the amount of audio channels (this one doesn't play on Chromecast, unlike the one above)
```
matt@thinkpod /mnt/3TB-Drive-01/Videos/Anime/Parasite Dolls $ mediainfo Parasite\ Dolls\ -\ Movie\ Release\ \[xxxxx\]\ \[DVD-480i\ \ MPEG-2\ AC-3\]\ \[9A796238\].mkv 
General
Unique ID                                : 181228527867187482418837538339673079100 (0x88575848B90F967AE4D758DF4383FD3C)
Complete name                            : Parasite Dolls - Movie Release [xxxx] [DVD-480i  MPEG-2 AC-3] [9A796238].mkv
Format                                   : Matroska
Format version                           : Version 2
File size                                : 4.22 GiB
Duration                                 : 1 h 25 min
Overall bit rate mode                    : Variable
Overall bit rate                         : 7 044 kb/s
Encoded date                             : UTC 2015-10-26 05:12:39
Writing application                      : MakeMKV v1.9.7 win(x86-release)
Writing library                          : libmakemkv v1.9.7 (1.3.0/1.4.1) win(x86-release)

Video
ID                                       : 1
Format                                   : MPEG Video
Format version                           : Version 2
Format profile                           : Main@Main
Format settings                          : CustomMatrix / BVOP
Format settings, BVOP                    : Yes
Format settings, Matrix                  : Custom
Format settings, GOP                     : M=3, N=12
Format settings, picture structure       : Frame
Codec ID                                 : V_MPEG2
Codec ID/Info                            : MPEG 1 or 2 Video
Duration                                 : 1 h 25 min
Bit rate mode                            : Variable
Bit rate                                 : 6 456 kb/s
Maximum bit rate                         : 8 900 kb/s
Width                                    : 720 pixels
Height                                   : 480 pixels
Display aspect ratio                     : 4:3
Frame rate mode                          : Constant
Frame rate                               : 29.970 (30000/1001) FPS
Standard                                 : NTSC
Color space                              : YUV
Chroma subsampling                       : 4:2:0
Bit depth                                : 8 bits
Scan type                                : Interlaced
Scan order                               : Top Field First
Compression mode                         : Lossy
Bits/(Pixel*Frame)                       : 0.623
Time code of first frame                 : 01:00:00:00
Time code source                         : Group of pictures header
GOP, Open/Closed                         : Open
GOP, Open/Closed of first frame          : Closed
Stream size                              : 3.87 GiB (92%)
Language                                 : English
Default                                  : No
Forced                                   : No

Audio
ID                                       : 2
Format                                   : AC-3
Format/Info                              : Audio Coding 3
Commercial name                          : Dolby Digital
Codec ID                                 : A_AC3
Duration                                 : 1 h 25 min
Bit rate mode                            : Constant
Bit rate                                 : 448 kb/s
Channel(s)                               : 6 channels
Channel layout                           : L R C LFE Ls Rs
Sampling rate                            : 48.0 kHz
Frame rate                               : 31.250 FPS (1536 SPF)
Bit depth                                : 16 bits
Compression mode                         : Lossy
Stream size                              : 275 MiB (6%)
Title                                    : Surround 5.1
Language                                 : English
Service kind                             : Complete Main
Default                                  : Yes
Forced                                   : No

Text
ID                                       : 3
Format                                   : VobSub
Codec ID                                 : S_VOBSUB
Codec ID/Info                            : Picture based subtitle format used on DVDs
Language                                 : English
Default                                  : Yes
Forced                                   : No

Menu
00:00:00.066                             : en:Chapter 01
00:01:31.357                             : en:Chapter 02
00:10:31.097                             : en:Chapter 03
00:21:24.549                             : en:Chapter 04
00:26:33.658                             : en:Chapter 05
00:33:12.557                             : en:Chapter 06
00:41:33.057                             : en:Chapter 07
00:51:47.771                             : en:Chapter 08
01:02:08.357                             : en:Chapter 09
01:09:10.379                             : en:Chapter 10
01:21:46.434                             : en:Chapter 11
01:25:43.271                             : en:Chapter 12
```

Another example of a non working file (transcoded process for the unsupported audio codec and audio channel setup never gets kicked off)
```
matt@thinkpod /mnt/1TB-Drive-02/Videos/TV Shows/Doctor Who (2005)/Specials $ mediainfo Doctor.Who.2005.Twice.Upon.a.Time.1080i.BluRay.REMUX.AVC.DTS-HD.MA.5.1-EPSiLON.mkv 
General
Unique ID                                : 129185394686927152346116505537618229683 (0x613033B64E150C79726FD4F7EB41B9B3)
Complete name                            : Doctor.Who.2005.Twice.Upon.a.Time.1080i.BluRay.REMUX.AVC.DTS-HD.MA.5.1-EPSiLON.mkv
Format                                   : Matroska
Format version                           : Version 4
File size                                : 13.3 GiB
Duration                                 : 59 min 54 s
Overall bit rate mode                    : Variable
Overall bit rate                         : 31.8 Mb/s
Encoded date                             : UTC 2018-02-13 07:22:51
Writing application                      : mkvmerge v20.0.0 ('I Am The Sun') 64-bit
Writing library                          : libebml v1.3.5 + libmatroska v1.4.8

Video
ID                                       : 1
Format                                   : AVC
Format/Info                              : Advanced Video Codec
Format profile                           : High@L4.1
Format settings                          : CABAC / 4 Ref Frames
Format settings, CABAC                   : Yes
Format settings, ReFrames                : 4 frames
Codec ID                                 : V_MPEG4/ISO/AVC
Duration                                 : 59 min 54 s
Bit rate mode                            : Variable
Bit rate                                 : 29.8 Mb/s
Maximum bit rate                         : 33.0 Mb/s
Width                                    : 1 920 pixels
Height                                   : 1 080 pixels
Display aspect ratio                     : 16:9
Frame rate mode                          : Constant
Frame rate                               : 25.000 FPS
Color space                              : YUV
Chroma subsampling                       : 4:2:0
Bit depth                                : 8 bits
Scan type                                : MBAFF
Scan type, store method                  : Interleaved fields
Scan order                               : Top Field First
Bits/(Pixel*Frame)                       : 0.574
Stream size                              : 12.5 GiB (94%)
Title                                    : MPEG-4 AVC Video / 29995 kbps / 1080i / 25 fps / 16:9 / High Profile 4.1
Language                                 : English
Default                                  : Yes
Forced                                   : No

Audio
ID                                       : 2
Format                                   : DTS XLL
Format/Info                              : Digital Theater Systems
Commercial name                          : DTS-HD Master Audio
Codec ID                                 : A_DTS
Duration                                 : 59 min 54 s
Bit rate mode                            : Variable
Bit rate                                 : 1 984 kb/s
Channel(s)                               : 6 channels
Channel layout                           : C L R Ls Rs LFE
Sampling rate                            : 48.0 kHz
Frame rate                               : 93.750 FPS (512 SPF)
Bit depth                                : 16 bits
Compression mode                         : Lossless
Stream size                              : 850 MiB (6%)
Title                                    : DTS-HD Master Audio / 5.1 / 48 kHz / 1984 kbps / 16-bit
Language                                 : English
Default                                  : Yes
Forced                                   : No

Text
ID                                       : 3
Format                                   : PGS
Codec ID                                 : S_HDMV/PGS
Codec ID/Info                            : Picture based subtitle format used on BDs/HD-DVDs
Duration                                 : 59 min 22 s
Bit rate                                 : 40.5 kb/s
Count of elements                        : 1924
Stream size                              : 17.2 MiB (0%)
Title                                    : SDH
Language                                 : English
Default                                  : No
Forced                                   : No

Menu
00:00:00.000                             : en:Chapter 1
00:04:40.160                             : en:Chapter 2
00:09:30.240                             : en:Chapter 3
00:14:06.440                             : en:Chapter 4
00:17:25.920                             : en:Chapter 5
00:26:53.560                             : en:Chapter 6
00:33:33.280                             : en:Chapter 7
00:39:23.560                             : en:Chapter 8
00:42:58.000                             : en:Chapter 9
00:50:02.960                             : en:Chapter 10
00:54:30.920                             : en:Chapter 11
00:59:18.800                             : en:Chapter 12
```
Now let's say I just change the audio codec, but leave the amount of audio channels the same, then the ffmpeg process will kick, and start transcoding, thus I am able to play it on the chrome cast.

```
matt@thinkpod ~/Videos/Doctor Who (2005)/Specials $ mediainfo test.mkv 
General
Unique ID                                : 147499026518799494604785278600885918066 (0x6EF7477942EBD6D17BC84BCC0F355972)
Complete name                            : test.mkv
Format                                   : Matroska
Format version                           : Version 4
File size                                : 1.45 GiB
Duration                                 : 7 min 20 s
Overall bit rate mode                    : Variable
Overall bit rate                         : 28.4 Mb/s
Writing application                      : Lavf58.20.100
Writing library                          : Lavf58.20.100
ErrorDetectionType                       : Per level 1

Video
ID                                       : 1
Format                                   : AVC
Format/Info                              : Advanced Video Codec
Format profile                           : High@L4.1
Format settings                          : CABAC / 4 Ref Frames
Format settings, CABAC                   : Yes
Format settings, ReFrames                : 4 frames
Codec ID                                 : V_MPEG4/ISO/AVC
Duration                                 : 7 min 20 s
Bit rate mode                            : Variable
Bit rate                                 : 29.8 Mb/s
Maximum bit rate                         : 33.0 Mb/s
Width                                    : 1 920 pixels
Height                                   : 1 080 pixels
Display aspect ratio                     : 16:9
Frame rate mode                          : Variable
Frame rate                               : 204.121 FPS
Original frame rate                      : 25.000 FPS
Color space                              : YUV
Chroma subsampling                       : 4:2:0
Bit depth                                : 8 bits
Scan type                                : MBAFF
Scan type, store method                  : Interleaved fields
Scan order                               : Top Field First
Bits/(Pixel*Frame)                       : 0.070
Stream size                              : 12.5 GiB
Title                                    : MPEG-4 AVC Video / 29995 kbps / 1080i / 25 fps / 16:9 / High Profile 4.1
Language                                 : English
Default                                  : Yes
Forced                                   : No

Audio
ID                                       : 2
Format                                   : AAC LC
Format/Info                              : Advanced Audio Codec Low Complexity
Codec ID                                 : A_AAC-2
Duration                                 : 7 min 20 s
Bit rate                                 : 1 984 kb/s
Channel(s)                               : 6 channels
Channel layout                           : L R C Cb Rls Rb
Sampling rate                            : 48.0 kHz
Frame rate                               : 46.875 FPS (1024 SPF)
Compression mode                         : Lossy
Stream size                              : 850 MiB (57%)
Title                                    : DTS-HD Master Audio / 5.1 / 48 kHz / 1984 kbps / 16-bit
Writing library                          : Lavc58.35.100 aac
Language                                 : English
Default                                  : Yes
Forced                                   : No

Menu
00:00:00.000                             : :Chapter 1
00:04:40.160                             : :Chapter 2
00:09:30.240                             : :Chapter 3
00:14:06.440                             : :Chapter 4
00:17:25.920                             : :Chapter 5
00:26:53.560                             : :Chapter 6
00:33:33.280                             : :Chapter 7
00:39:23.560                             : :Chapter 8
00:42:58.000                             : :Chapter 9
00:50:02.960                             : :Chapter 10
00:54:30.920                             : :Chapter 11
00:59:18.800                             : :Chapter 12
```

**Logs Expected behavior**

Heres what happens when I play the test file (unsupported audio channel configuration for the Chromecast)
```
[18:19:13] [INF] User policy for matt. EnablePlaybackRemuxing: True EnableVideoPlaybackTranscoding: True EnableAudioPlaybackTranscoding: True
[18:19:13] [INF] Profile: Unknown Profile, Path: /home/matt/Videos/Doctor Who (2005)/Specials/test.mkv, isEligibleForDirectPlay: True, isEligibleForDirectStream: True
[18:19:13] [INF] Profile: VideoAudioCodecProfile, DirectPlay=false. Reason=Unknown Profile.AudioChannels Condition: LessThanEqual. ConditionValue: 2. IsRequired: True. Path: /home/matt/Videos/Doctor Who (2005)/Specials/test.mkv
[18:19:13] [INF] Profile: Unknown Profile, Path: /home/matt/Videos/Doctor Who (2005)/Specials/test.mkv, isEligibleForDirectPlay: True, isEligibleForDirectStream: True
[18:19:13] [INF] Profile: VideoAudioCodecProfile, DirectPlay=false. Reason=Unknown Profile.AudioChannels Condition: LessThanEqual. ConditionValue: 2. IsRequired: True. Path: /home/matt/Videos/Doctor Who (2005)/Specials/test.mkv
[18:19:13] [INF] Profile: Unknown Profile, Path: /home/matt/Videos/Doctor Who (2005)/Specials/test.mkv, isEligibleForDirectPlay: True, isEligibleForDirectStream: True
[18:19:13] [INF] Profile: VideoAudioCodecProfile, DirectPlay=false. Reason=Unknown Profile.AudioChannels Condition: LessThanEqual. ConditionValue: 2. IsRequired: True. Path: /home/matt/Videos/Doctor Who (2005)/Specials/test.mkv
[18:19:13] [INF] Playback Started
[18:19:13] [INF] Could not match user with any stored credentials
[18:19:13] [INF] /usr/bin/ffmpeg -fflags +genpts -f matroska,webm -i file:"/home/matt/Videos/Doctor Who (2005)/Specials/test.mkv" -map 0:0 -map 0:1 -map -0:s -codec:v:0 copy -copyts -avoid_negative_ts disabled -start_at_zero -vsync -1 -map_metadata -1 -map_chapters -1 -threads 0 -codec:a:0 aac -strict experimental -ac 2 -ab 320000 -af "volume=2" -y "/home/matt/.local/share/jellyfin/transcoding-temp/0d9c9a2f9c239d46c58b96e461f68ab0.mkv"
[18:19:32] [INF] FFMpeg exited with code 0
```
Heres what happens when I play the original (again only difference is audio codec, so basically unsupported audio codec and audio channel configuration this time.)
```
[18:22:00] [INF] User policy for matt. EnablePlaybackRemuxing: True EnableVideoPlaybackTranscoding: True EnableAudioPlaybackTranscoding: True
[18:22:00] [INF] Profile: Unknown Profile, Path: /mnt/1TB-Drive-02/Videos/TV Shows/Doctor Who (2005)/Specials/Doctor.Who.2005.Twice.Upon.a.Time.1080i.BluRay.REMUX.AVC.DTS-HD.MA.5.1-EPSiLON.mkv, isEligibleForDirectPlay: True, isEligibleForDirectStream: True
[18:22:00] [INF] Profile: Unknown Profile, No direct play profiles found for Path: /mnt/1TB-Drive-02/Videos/TV Shows/Doctor Who (2005)/Specials/Doctor.Who.2005.Twice.Upon.a.Time.1080i.BluRay.REMUX.AVC.DTS-HD.MA.5.1-EPSiLON.mkv
[18:22:00] [INF] Profile: Unknown Profile, Path: /mnt/1TB-Drive-02/Videos/TV Shows/Doctor Who (2005)/Specials/Doctor.Who.2005.Twice.Upon.a.Time.1080i.BluRay.REMUX.AVC.DTS-HD.MA.5.1-EPSiLON.mkv, isEligibleForDirectPlay: True, isEligibleForDirectStream: True
[18:22:00] [INF] Profile: Unknown Profile, No direct play profiles found for Path: /mnt/1TB-Drive-02/Videos/TV Shows/Doctor Who (2005)/Specials/Doctor.Who.2005.Twice.Upon.a.Time.1080i.BluRay.REMUX.AVC.DTS-HD.MA.5.1-EPSiLON.mkv
[18:22:00] [INF] Profile: Unknown Profile, Path: /mnt/1TB-Drive-02/Videos/TV Shows/Doctor Who (2005)/Specials/Doctor.Who.2005.Twice.Upon.a.Time.1080i.BluRay.REMUX.AVC.DTS-HD.MA.5.1-EPSiLON.mkv, isEligibleForDirectPlay: True, isEligibleForDirectStream: True
[18:22:00] [INF] Profile: Unknown Profile, No direct play profiles found for Path: /mnt/1TB-Drive-02/Videos/TV Shows/Doctor Who (2005)/Specials/Doctor.Who.2005.Twice.Upon.a.Time.1080i.BluRay.REMUX.AVC.DTS-HD.MA.5.1-EPSiLON.mkv
```
The line that catches my eye is this
```
[18:19:13] [INF] Profile: VideoAudioCodecProfile, DirectPlay=false. Reason=Unknown Profile.AudioChannels Condition: LessThanEqual. ConditionValue: 2. IsRequired: True. Path: /home/matt/Videos/Doctor Who (2005)/Specials/test.mkv
```
As you can clearly see, jellyfin realises this is an unsupported setup for the Chromecast so it transcodes it to something playable, this line isn't visible in the log for the second example.

**System (please complete the following information):**
 - OS: Solus (note: ffmpeg version 4.1, but this above problem is also happening for others on different setups, so I don't believe ffmpeg itself is to blame)
 - Browser: Chromecast
 - Jellyfin Version: 10.3.1
 - Reverse proxy: NA

**Additional context**
Basically what I believe it boils down to is unsupported audio codec and/or audio channel setups aren't playable on the Chromecast, and the check for this isn't always happening in jellyfin thats why some files are playable and some aren't.
Note: I am using a Chromecast 3rd gen, but the Reddit thread link suggests it isn't a problem specific to the 3rd Gen Chromecast

**Comment by cvium** [comment]
Do you have a test file?

**Comment by anthonylavado** [comment]
While I don’t know that any of these in particular match, I’m sure something from either of these sites would suffice:

https://kodi.wiki/view/Samples

http://jell.yfish.us/

**Comment by brianjmurrell** [comment]
This looks like what I am seeing but I can reproduce it without even chromecasting.  A certain 2 channel audio file plays, remuxed with ffmpeg to chrome on linux with a 2-speaker output and the same file fails to play on chrome on my phone.  The obvious difference in the logs is that when it works, I see the line indicating that an ffmpeg was started and when it fails to play on my phone, there is no ffmpeg starting log entry.

**Comment by Matt07211** [comment]
Something I've also noticed, this problem seems to be occurring on any Chromecast that isn't the big standard gen 2 defined within jellyfin. E.g. th current gen 3 Chromecasts, or e.g.2. gen 2 cc that are embedded into a tv (re: Reddit link)

**Comment by Matt07211** [comment]
Finally found where the audio channels and audio codec support is listed/checked

https://github.com/jellyfin/jellyfin-web/blob/82f171e99a705b6deb46805ff2332363cfe23f7b/src/components/browserdeviceprofile.js#L758

**Comment by Matt07211** [comment]
So testing off latest 10.3.2 this some how fixed itself, even though this was a problem on 10.3.0 release and 10.3.1 on git (after the fix that fixed seeking)

**Comment by brianjmurrell** [comment]
This is not fixed for me in 10.3.3

**Comment by Matt07211** [comment]
Have you restarted the Chromecast?

I'll see what commit I compiled my 10.3.3 version and see if that works for you. If it does all we need to do is got bisect

**Comment by brianjmurrell** [comment]
I probably can't do anything until #1396 is figured out as the web-client doesn't even seem to be able to initiate playback on Chromecast any more.  From the complete lack of activity in either the jellyfin log nor the browser devtools, it would seem that Chromecast is not even involved yet.
